crates/project/src/lsp_command.rs

```diff
@@ -3178,3 +3183,477 @@ impl LspCommand for LinkedEditingRange {
+    type Response = Option<LspDiagnostics>;
+    type LspRequest = lsp::request::DocumentDiagnosticRequest;
+    type ProtoRequest = proto::GetDocumentDiagnostics;
+
+    fn display_name(&self) -> &str {
+        "Get diagnostics"
+    }
+
+    fn check_capabilities(&self, server_capabilities: AdapterServerCapabilities) -> bool {
+        server_capabilities
+            .server_capabilities
+            .diagnostic_provider
+            .is_some()
+    }
+
+    fn to_lsp(
+        &self,
+        path: &Path,
+        _: &Buffer,
+        language_server: &Arc<LanguageServer>,
+        _: &AppContext,
+    ) -> Result<lsp::DocumentDiagnosticParams> {
+        let identifier = match language_server.capabilities().diagnostic_provider {
+            Some(lsp::DiagnosticServerCapabilities::Options(options)) => options.identifier.clone(),
+            Some(lsp::DiagnosticServerCapabilities::RegistrationOptions(options)) => {
+                options.diagnostic_options.identifier.clone()
+            }
+            None => None,
+        };
+
+        Ok(lsp::DocumentDiagnosticParams {
+            text_document: lsp::TextDocumentIdentifier {
+                uri: lsp::Url::from_file_path(path).unwrap(),
+            },
+            identifier,
+            previous_result_id: None,
+            partial_result_params: Default::default(),
+            work_done_progress_params: Default::default(),
+        })
+    }
+
+    async fn response_from_lsp(
+        self,
+        message: lsp::DocumentDiagnosticReportResult,
+        lsp_store: Model<LspStore>,
+        buffer: Model<Buffer>,
+        server_id: LanguageServerId,
+        mut cx: AsyncAppContext,
+    ) -> Result<Self::Response> {
+        let uri = buffer.read_with(&cx, |buffer, cx| {
+            buffer
+                .file()
+                .and_then(|file| file.as_local())
+                .map(|file| lsp::Url::from_file_path(file.abs_path(cx).clone()).unwrap())
+        })?;
+
+        let Some(uri) = uri else {
+            return Ok(None);
+        };
+
+        match message {
+            lsp::DocumentDiagnosticReportResult::Report(report) => match report {
+                lsp::DocumentDiagnosticReport::Full(report) => {
+                    lsp_store.update(&mut cx, |store, cx| {
+                        for (uri, report) in report.related_documents.into_iter().flatten() {
+                            match report {
+                                lsp::DocumentDiagnosticReportKind::Full(report) => {
+                                    store.update_diagnostics(
+                                        server_id,
+                                        lsp::PublishDiagnosticsParams {
+                                            diagnostics: report.items.clone(),
+                                            uri,
+                                            version: None,
+                                        },
+                                        &[],
```

Comment:
How true is this?
There's a special code path for rust-analyzer's diagnostics that used to check whether the diagnostics pushed came from `cargo` or from r-a, is it so that we're losing this info now?

Reply:
Hmm, yeah, you are correct. For Rust, there is `rustc` source - https://github.com/vitallium/zed/blob/17db3cd154e733edb666c468a273a69941b28ac1/crates/languages/src/rust.rs#L220 I don't think there is a way to get disk based sources inside the `lsp_command` so probably there should be another method in `lsp_store` that combines pulled diagnostics and disk-based sources. I will check that.

Reply:
Actually, I found that it's possible to this:

```rust
let language_server_adapter = lsp_store
            .update(&mut cx, |lsp_store, _| {
                lsp_store.language_server_adapter_for_id(server_id)
            })?
            .ok_or_else(|| anyhow!("no such language server"))?;
````

in the `response_from_lsp` method to get LSP adapter which has the method `disk_based_diagnostic_sources` that is required here.